docs/recipes/pipelines.ipynb

```diff
@@ -709,7 +699,282 @@
+    "\n",
+    "One may wonder what is the advantage of learning during predict. Let's compare the performance of a pipeline with and without learning during predict, in two scenarios: one in which the flow of data stays the same, we just update "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from contextlib import nullcontext\n",
+    "from river import metrics\n",
+    "\n",
+    "import pandas as pd"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def score_pipeline(learn_during_predict: bool, n_learning_samples: int | None = None) -> float:\n",
+    "    \"\"\"Scores a pipeline on the TrumpApproval dataset.\n",
+    "\n",
+    "    Parameters\n",
+    "    ----------\n",
+    "    learn_during_predict : bool\n",
+    "        Whether or not to learn the unsupervided components during the prediction step.\n",
+    "        If False it will only learn when `learn_one` is explicitly called.\n",
+    "    n_learning_samples : int | None \n",
+    "        Number of samples used to `learn_one`.\n",
+    "\n",
+    "    Return\n",
+    "    ------\n",
+    "    MAE : float\n",
+    "        Mean absolute error of the pipeline on the dataset\n",
+    "    \"\"\"\n",
+    "\n",
+    "    dataset = datasets.TrumpApproval()\n",
+    "\n",
+    "    model = (\n",
+    "        preprocessing.StandardScaler() |\n",
+    "        linear_model.LinearRegression()\n",
+    "        )\n",
+    "\n",
+    "    metric = metrics.MAE()\n",
+    "\n",
+    "    ctx = compose.learn_during_predict if learn_during_predict else nullcontext\n",
+    "    n_learning_samples = n_learning_samples or dataset.n_samples\n",
+    "\n",
+    "    with ctx():\n",
+    "        for _idx, (x, y) in enumerate(dataset):\n",
+    "            y_pred = model.predict_one(x)\n",
+    "\n",
+    "            metric.update(y, y_pred)\n",
+    "            \n",
+    "            if _idx < n_learning_samples:\n",
+    "                model.learn_one(x, y)\n",
+    "\n",
+    "    return metric.get()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "max_samples = datasets.TrumpApproval().n_samples\n",
+    "\n",
+    "results = [\n",
+    "    {\n",
+    "        \"learn_during_predict\": learn_during_predict,\n",
+    "        \"pct_learning_samples\": round(100*n_learning_samples/max_samples, 0),\n",
+    "        \"mae\": score_pipeline(learn_during_predict=learn_during_predict, n_learning_samples=n_learning_samples)\n",
+    "    }\n",
+    "    for learn_during_predict in (True, False)\n",
+    "    for n_learning_samples in range(max_samples, max_samples//10, -(max_samples//10))\n",
+    "]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
```

Comment:
Best way to see this is to spin the docs locally or wherever.
Here a snapshot of how it renders in my local:

![image](https://github.com/online-ml/river/assets/42817048/f185e0c5-cab1-42e6-a5e2-37b730c0a303)

Reply:
I got to admit, I was not expecting this much of a difference! That's awesome

Reply:
> I got to admit, I was not expecting this much of a difference! That's awesome.

Yeah and you're actually learning from observations directly after predicting them (before seeing the next sample). In real world settings there is usually a lag before the ground truth gets available, which makes you predict for way more samples before you can learn from complete ones. A more realistic benchmark would probably be even more impressive.

> One question I have however is: does the unsupervised part learn from the same datapoint twice? i.e. both during `predict_*` and `learn_*`?

Nope, the context manager sets the [`_LEARN_UNSUPERVISED_DURING_PREDICT` var to `True`](https://github.com/online-ml/river/blob/c7fbedfaba9d68622588b79b23cca20a115e5d2d/river/compose/pipeline.py#L114) so that the unsupervised transformers do not get updated within `learn_*` calls (see [condition](https://github.com/online-ml/river/blob/c7fbedfaba9d68622588b79b23cca20a115e5d2d/river/compose/pipeline.py#L448-L450))